enhancements/20231108-elemental-cloudinit-crd.md

```diff
@@ -0,0 +1,376 @@
+      Reason:                CloudInitPresentOnDisk
+      Status:                True
+      Type:                  Present
+    Node Name:               harv1
+```
+
+In the case of this single-node cluster, the user can see that the
+cloud-init-style file has been copied down to `/oem/99_ssh.yaml` on
+node `harv1`.
+
+They can also see that the file is up-to-date since the rollout says
+it is not `OutOfSync`.
+
+The user probably already knows that it is `Applicable` to `harv` because
+the `ssh-access` CloudInit resource spec says `matchSelector: {}`, which is
+a label set that will match every node. That said, if they updated the
+`matchSelector` down the line, then the `Applicable` condition may help
+them troubleshoot why a node has or does not have a cloud-init-style file
+copied down to it.
+
+The user can now reboot the Harvester node so that their cloud-init-style
+file may be applied at boot time.
+
+#### Add/rotate rancher user authorized keys (UI/Dashboard)
+
+Alternatively, instead of creating the `ssh-access` resource via kubectl,
+there could be a new form added to the Harvester dashboard (perhaps as a
+sub-page under the "Hosts" section.) This page could have a table with
+a list of the cluster's CloudInit resources.
+
+At the far right of each row for each CloudInit resource, there could be a
+3-dot menu much like the Virtual Machines page (and others), where a Harvester
+user could click "Edit Config", "Edit YAML", or "delete."
+
+There could also be a button called "Create".
+
+The "Create" and "Edit CloudInit" pages could have a post form that looks
+a little something like this (though the "Edit CloudInit page" will have
+values pre-populated such as resource name, file name, labels, contents):
+
+1. CloudInit resource name (in the case of the above example, `ssh-access`.)
+1. The base filename (in the above example, `99_ssh.yaml`.)
+1. A `key` input and a `value` input (like the `Labels` form while editing
+a Host config). Another row of `key` and `value` inputs can be added by clicking
+the `Add Label` button. In the case of the `ssh-access` example above, these
+would be left empty so that the UI passes an empty match selector to match all
+nodes (`matchSelector: {}`.)
+1. A textbox where the cloud-init YAML contents go.
+1. A "save" button (or it could be labeled "Create" or "Update" depending on
+   if this is a new CloudInit object or if the user is editing a pre-existing one.)
+
+After clicking "save/create/update", the dashboard could redirect the user to
+a more detailed page (much like the one you see for a virtual machine when clicking
+on a virtual machine's name on the "Virtual Machines" page of the dashboard.)
+
+This page could present the object's spec and its status fields in a way that
+the user can understand the state of the `ssh-access` rollout as described above
+with the `kubectl` example.
+
+After the user is satisfied with the CloudInit object they've created,
```

Comment:
Should we consider to add a 'Reboot' button when a new CloudInit is created/updated ?

Reply:
This could be cool but I don't think the juice is worth the squeeze since there are some special cases that might require a disproportional amount of effort to deliver this.

For example, if someone makes a change to the document's formatting or ordering of fields, but the end result's semantic meaning is identical, like this:

```yaml
a: foo
b: bar
```

```yaml
b: bar
a: foo
```

then we'd want to avoid the confusion of showing a reboot prompt on the dashboard as it's not necessary, but the file on disk would be updated due to the change regardless, so in this situation, a changed file is not always a good signal for whether a reboot is required.

Furthermore, I don't think the backend has node power management capabilities (as of yet) and that would probably be out of scope of the node-manager component that this HEP is targeting.